src/watchdog.rs

```diff
@@ -0,0 +1,210 @@
+                trigger_panic: value.panic_on_trigger,
+                idle_core_mask: value
+                    .subscribed_idle_tasks
+                    .iter()
+                    .fold(0u32, |mask, core| mask & (1 << (*core as u32))),
+            }
+        }
+    }
+
+    pub trait Twdt {}
+
+    pub struct TWDTDriver<'d>(PhantomData<&'d mut ()>);
+
+    unsafe impl Send for TWDTDriver<'_> {}
+
+    static TWDT_DRIVER_REF_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    impl Clone for TWDTDriver<'_> {
+        fn clone(&self) -> Self {
+            TWDT_DRIVER_REF_COUNT.fetch_add(1, core::sync::atomic::Ordering::Relaxed);
+            Self(Default::default())
+        }
+    }
+
+    impl<'d> TWDTDriver<'d> {
+        pub fn new(
+            _twdt: impl Peripheral<P = TWDT> + 'd,
+            config: &WatchdogConfig,
+        ) -> Result<Self, EspError> {
+            init_or_reconfigure(config)?;
+            Ok(Self(Default::default()))
+        }
+
+        // will fail in isr
+        pub fn watch_current_task(&mut self) -> Result<WatchdogSubscription<'_>, EspError> {
+            add_task(core::ptr::null_mut())?;
+            Ok(WatchdogSubscription::new())
+        }
+    }
+
+    impl Drop for TWDTDriver<'_> {
+        fn drop(&mut self) {
+            let _drop = TWDT_DRIVER_REF_COUNT
+                .fetch_update(
+                    core::sync::atomic::Ordering::Relaxed,
+                    core::sync::atomic::Ordering::Relaxed,
+                    |count| if count == 0 { None } else { Some(count - 1) },
+                )
+                .is_err();
+            // TODO: maybe deinit twdt from idf (only if not initialized by idf)
+        }
+    }
+
+    pub struct WatchdogSubscription<'s>(PhantomData<&'s mut ()>);
+
+    impl WatchdogSubscription<'_> {
+        fn new() -> Self {
+            Self(Default::default())
+        }
+    }
+
+    impl embedded_hal_0_2::watchdog::Watchdog for WatchdogSubscription<'_> {
+        fn feed(&mut self) {
+            feed_from_curr_task().unwrap()
+        }
+    }
+
+    impl Drop for WatchdogSubscription<'_> {
+        fn drop(&mut self) {
+            let _ = remove_task(core::ptr::null_mut());
+        }
+    }
+
+    crate::impl_peripheral!(TWDT);
+    impl Twdt for TWDT {}
+
+    fn init_or_reconfigure(config: &WatchdogConfig) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_task_wdt_init(
+                config.duration.as_secs() as u32,
+                config.panic_on_trigger
+            ))
+        }
+    }
+
+    fn feed_from_curr_task() -> Result<(), EspError> {
+        unsafe { esp!(esp_task_wdt_reset()) }
+    }
+
+    fn add_task(task: TaskHandle_t) -> Result<(), EspError> {
+        unsafe { esp!(esp_task_wdt_add(task)) }
+    }
+
+    fn remove_task(task: TaskHandle_t) -> Result<(), EspError> {
+        match unsafe { esp_task_wdt_delete(task) } {
+            ESP_ERR_INVALID_ARG | ESP_OK => Ok(()),
```

Comment:
Why getting `ESP_ERR_INVALID_TASK` is OK? What happens if I try to watch a task twice? Hopefully, ESP IDF returns an error on the second attempt anyway?

Reply:
If `ESP_ERR_INVALID_TASK` should not happen, it is even more important **not** to swallow it silently, as if it happens then this is a bug which should not be kept hidden. Just do the `esp!(...)?` dance as usu.